Detectors/TPC/calibration/src/CalibdEdx.cxx

```diff
@@ -0,0 +1,104 @@
+
+#include "TPCCalibration/CalibdEdx.h"
+
+#include <array>
+#include <cstddef>
+#include <memory>
+#include <string_view>
+
+//o2 includes
+#include "CommonUtils/MemFileHelper.h"
+#include "CCDB/CcdbApi.h"
+#include "DetectorsCalibration/Utils.h"
+#include "Framework/Logger.h"
+#include "TPCCalibration/dEdxHistos.h"
+
+//root includes
+#include "TFile.h"
+#include "TTree.h"
+
+using namespace o2::tpc;
+
+void CalibdEdx::initOutput()
+{
+  // Here we initialize the vector of our output objects
+  mInfoVector.clear();
+  mMIPVector.clear();
+  return;
+}
+
+void CalibdEdx::finalizeSlot(Slot& slot)
+{
+  LOG(INFO) << "Finalizing slot " << slot.getTFStart() << " <= TF <= " << slot.getTFEnd();
+
+  const dEdxHistos* container = slot.getContainer();
+  const auto statsASide = container->getHist()[0].getStatisticsData();
+  const auto statsCSide = container->getHist()[1].getStatisticsData();
+
+  slot.print();
+  LOG(INFO) << "A side, truncated mean statistics: Mean = " << statsASide.mCOG << ", StdDev = " << statsCSide.mStdDev << ", Entries = " << statsASide.mSum;
+  LOG(INFO) << "C side, truncated mean statistics: Mean = " << statsCSide.mCOG << ", StdDev = " << statsCSide.mStdDev << ", Entries = " << statsCSide.mSum;
+
+  CalibMIP mips{statsASide.mCOG, statsCSide.mCOG};
+
+  const auto className = o2::utils::MemFileHelper::getClassName(mips);
+  const auto fileName = o2::ccdb::CcdbApi::generateFileName(className);
+  const std::map<std::string, std::string> metaData;
+
+  // TODO: the timestamp is now given with the TF index, but it will have
+  // to become an absolute time.
+  mInfoVector.emplace_back("TPC/Calib/MIPS", className, fileName, metaData, slot.getTFStart(), 99999999999999);
+  mMIPVector.push_back(mips);
+
+  if (mDumpStream) {
+    LOG(INFO) << "Dumping time slot data to file";
+    TFType tf = slot.getTFStart();
+    LOG(INFO) << "COG " << mips[0] << " " << mips[1];
+    *mDumpStream << "mipPosition"
+                 << "timeFrame=" << tf                   // Initial time frame of time slot
+                 << "calibMIP=" << mips                  // Computed MIP positions
+                 << "dEdxHistos=" << slot.getContainer() // dE/dx histograms
+                 << "\n";
+  }
+}
+
+CalibdEdx::Slot& CalibdEdx::emplaceNewSlot(bool front, TFType tstart, TFType tend)
+{
+  auto& cont = getSlots();
+  auto& slot = front ? cont.emplace_front(tstart, tend) : cont.emplace_back(tstart, tend);
+
+  auto container = std::make_unique<dEdxHistos>(mNBins, mCuts);
+  container->setApplyCuts(mApplyCuts);
+
+  slot.setContainer(std::move(container));
+  return slot;
+}
+
+void CalibdEdx::enableDumpToFile(std::string_view fileName)
+{
+  mDumpStream = std::make_unique<o2::utils::TreeStreamRedirector>(fileName.data(), "recreate");
+}
+
+void CalibdEdx::disableDumpToFile()
+{
+  // This will call the TreeStream destructor and write any stored data.
+  mDumpStream.reset();
+}
+
+void CalibdEdx::dumpToFile() const
```

Comment:
Also here the name is a bit misleading, since the dumping to file using the stream functionality is done in the place where it is used. Usually, dumpToFile should really put contents to a file. But this is not what is done here. So one might call it `finalizeDebugOutput` or so.